### PR TITLE
Implement general IoC container flow

### PR DESCRIPTION
**Pull request by nadya-gainutsa**
**Description:**

Added implementation of ApplicationContext (except `register` methods)

**Ticket:** [https://trello.com/c/xY6eziFf/23-implement-general-ioc-container-flow](https://trello.com/c/xY6eziFf/23-implement-general-ioc-container-flow)

### CODE REVIEW

**Comment by NazarSmith228** [comment]
1. No need to add Trello ticket number to PR title :)
2. Commit messages should be infinitive - implement instead of implemented, add instead of added etc.

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 1]
```diff
@@ -1,60 +1,152 @@
...
+
 
 /**
- * Example of ApplicationContext implementation, can be removed or reimplemented when actual development starts
+ * Implementation of ApplicationContext
  */
+@Slf4j
 public class ApplicationContextImpl implements ApplicationContext {
+    private Table<String, Class<?>, BeanDefinition> context = HashBasedTable.create();
+    private final String BEAN_TYPE_MUST_BE_NOT_NULL_MESSAGE = "The argument [beanType] must be not null";
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:23]: No need to extract these messages as constants. If they are used in multiple places, they should be fetched from somewhere - property file etc.  Use them directly in exception constructor for now.

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 2]
```diff
@@ -1,60 +1,152 @@
...
+import java.util.Objects;
+
 
 /**
- * Example of ApplicationContext implementation, can be removed or reimplemented when actual development starts
+ * Implementation of ApplicationContext
  */
+@Slf4j
 public class ApplicationContextImpl implements ApplicationContext {
+    private Table<String, Class<?>, BeanDefinition> context = HashBasedTable.create();
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: I would rename this field to `beanDefinitionTable` and remove direct instantiation

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 3]
```diff
@@ -1,60 +1,152 @@
...
                                   BeanValidator validator,
                                   BeanInitializer initializer) {
-        /* 1) Scanners get all beans
-           2) Merge List of bean definitions into one List
-           3) Send beans list to validator to check duplicates
-           4) Convert bean list to map
-           5) Send beans map to initializer to instantiate beans and inject dependencies
-         */
+
+        List<BeanDefinition> beanDefinitionList = scanningApplication(scanners);
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: 1. Usually method name should not be a noun. In most cases it should be a verb (e.g. `scanApplication()`)
2. Regarding method names - they should tell the reader what exactly method does - `scanApplication()` in this case will not work, since we're no scanning any application. We're just triggering `scan()` method in every `BeanScanner`. Thus I would name this method smth like `runScanners` (feel free to come up with a better name).
3. Now I would like to outline, that methods `validate()` and `beanInstantiation()` do nothing except for logging some message and delegate method call. Therefore, I would remove these since you can call them directly in constructor.
4. Method `contextCreation()` is also not correctly named, since you are just populating the container table with bean definitions. I would remove direct table instantiation from line 22 and change line 45 to `this.beanDefinitionTable = populateBeanDefinitionTable(beanDefinitionList)`. After that you can change line 46 to `initializer.initialize(beanDefinitionTable)`.

Having said that, I would change the constructor to look like this: 
```
List<BeanDefinition> beanDefinitionList = runScanners(scanners);
validator.validate(beanDefinitionList);
this.beanDefinitionTable = populateBeanDefinitionTable(beanDefinitionList);
initializer.initialize(beanDefinitionTable);
```
Feel free to contact me for discussion ;)

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 4]
```diff
@@ -1,60 +1,152 @@
...
+        beansInstantiation(initializer);
+
+        log.info("Application context initialized");
+    }
+
+    private List<BeanDefinition> scanningApplication(List<BeanScanner> scanners) {
+        List<BeanDefinition> beanDefinitionList = new ArrayList<>();
+        log.info("Scanning application for finding bean definitions");
+        for (BeanScanner beanScanner : scanners) {
+            log.debug("Running bean scanner {}", beanScanner.getClass().getSimpleName());
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: This log should be produced by the scanner itself - scanner should notify us that it started doing its job.

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 5]
```diff
@@ -1,60 +1,152 @@
...
+        } else if (beans.size() > 1) {
+            throw new NoUniqueBeanException("Expected single bean of type " + beanType.getSimpleName()
+                + ", but found " + beans.size());
+        } else return beans.get(0);
     }
 
     @Override
     public Object getBean(String beanName) {
-        return null;
+        Optional<BeanDefinition> beanDefinition = context.row(beanName)
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: I would rewrite it like: 
```
return context.row(beanName).values()
                .stream()
                .findFirst()
                .map(BeadDefinition::instance)
                .orElseThrow(() -> new NoSuchBeanException(...))
```

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 6]
```diff
@@ -1,60 +1,152 @@
...
+        }
     }
 
     @Override
     public <T> T getBean(String beanName, Class<T> beanType) {
-        return null;
+        Objects.requireNonNull(beanType, BEAN_TYPE_MUST_BE_NOT_NULL_MESSAGE);
+
+        Object bean = getBean(beanName);
+        if (bean.getClass().isAssignableFrom(beanType)) {
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:121]: (nit) We may simplify it by `if(beanType.isInstance(bean))`

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 7]
```diff
@@ -1,60 +1,152 @@
...
+        if (beans.isEmpty()) {
+            throw new NoSuchBeanException(NO_SUCH_BEAN_EXCEPTION_MESSAGE.formatted(beanType.getSimpleName()));
+        } else if (beans.size() > 1) {
+            throw new NoUniqueBeanException("Expected single bean of type " + beanType.getSimpleName()
+                + ", but found " + beans.size());
+        } else return beans.get(0);
     }
 
     @Override
     public Object getBean(String beanName) {
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:104]: We should check here that `beanName` is not null, empty or blank

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 8]
```diff
@@ -1,60 +1,152 @@
...
+                .findFirst();
+        if (beanDefinition.isPresent()) {
+            return beanDefinition.get().instance();
+        } else {
+            throw new NoSuchBeanException(NO_SUCH_BEAN_EXCEPTION_MESSAGE.formatted(beanName));
+        }
     }
 
     @Override
     public <T> T getBean(String beanName, Class<T> beanType) {
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:116]: We should also check here that `beanName` is not null, empty or blank

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 9]
```diff
@@ -1,60 +1,152 @@
...
+        Object bean = getBean(beanName);
+        if (bean.getClass().isAssignableFrom(beanType)) {
+            return beanType.cast(bean);
+        } else {
+            throw new NoSuchBeanException(NO_SUCH_BEAN_EXCEPTION_MESSAGE.formatted(beanName));
+        }
     }
 
     @Override
     public <T> Map<String, T> getAllBeans(Class<T> beanType) {
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:129]: We should check here that `beanType` is not null - just like in other methods you implemented :)

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 10]
```diff
@@ -1,60 +1,152 @@
...
-        return null;
+        return context.column(beanType)
+                .entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, entry -> beanType.cast(entry.getValue().instance())));
+
     }
 
     @Override
-    public boolean containsBean(String name) {
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:57]: We should check here that `beanName` is not null, empty or blank

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 11]
```diff
@@ -1,60 +1,152 @@
...
     @Override
     public void register(Class<?>... beanTypes) {
-
+        // TODO:  add implementation
     }
 
     @Override
     public <T> String register(Class<T> beanType) {
         return null;
+        // TODO: add implementation
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:83]: (nit) It is better to insert `TODO` before return statement

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 12]
```diff
@@ -1,60 +1,152 @@
...
+
+        List<T> beans = context.column(beanType)
+                .values()
+                .stream()
+                .map(beanDefinition -> beanType.cast(beanDefinition.instance()))
+                .toList();
+        if (beans.isEmpty()) {
+            throw new NoSuchBeanException(NO_SUCH_BEAN_EXCEPTION_MESSAGE.formatted(beanType.getSimpleName()));
+        } else if (beans.size() > 1) {
+            throw new NoUniqueBeanException("Expected single bean of type " + beanType.getSimpleName()
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: You can use `format()` here - `String.format("Found %d instances of bean %s",  beans.size(), beanType.getSimpleName())`

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 13]
```diff
@@ -1,60 +1,152 @@
...
+        validate(validator, beanDefinitionList);
+        contextCreation(beanDefinitionList);
+        beansInstantiation(initializer);
+
+        log.info("Application context initialized");
+    }
+
+    private List<BeanDefinition> scanningApplication(List<BeanScanner> scanners) {
+        List<BeanDefinition> beanDefinitionList = new ArrayList<>();
+        log.info("Scanning application for finding bean definitions");
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: I would rephrase this log message to `Started bean definition scanning`

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 14]
```diff
@@ -1,60 +1,152 @@
...
+        return beanDefinitionList;
+    }
+
+    private void validate(BeanValidator validator, List<BeanDefinition> beanDefinitionList) {
+        log.info("Beans definitions validating");
+        validator.validate(beanDefinitionList);
+    }
+
+    private void contextCreation(List<BeanDefinition> beanDefinitionList) {
+        log.debug("Context table generation");
```
Comment (NazarSmith228) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: I would rephrase this log message to `Store scanned bean definitions in application context`

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 7 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:104]: I agree with you about null check. But we discussed at the meeting that empty or blank name is acceptable.

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 9 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:129]: Oh yes, of course. It's obvious :)

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 3 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: Thank you for the detailed explanation. Honestly, this piece of code doesn't look good to me neither. But I didn't know what to change.

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 7 above)
Comment (NazarSmith228) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:104]: @nadya-gainutsa uh, okay, will discuss it with others...

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 15]
```diff
@@ -1,60 +1,152 @@
...
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.Objects;
+
 
 /**
- * Example of ApplicationContext implementation, can be removed or reimplemented when actual development starts
+ * Implementation of ApplicationContext
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: I think we should either remove or extend this comment somehow

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 16]
```diff
@@ -1,60 +1,152 @@
...
  */
+@Slf4j
 public class ApplicationContextImpl implements ApplicationContext {
+    private Table<String, Class<?>, BeanDefinition> context = HashBasedTable.create();
+    private final String BEAN_TYPE_MUST_BE_NOT_NULL_MESSAGE = "The argument [beanType] must be not null";
+    private final String NO_SUCH_BEAN_EXCEPTION_MESSAGE = "Bean with provided name/type [%s] not found in the context";
+
+    /**
+     *  Scanners scan application to define beans configs and map them into {@link BeanDefinition} lists.
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: I'd update it to "Scanners scan application _**packages**_..." 
Also, I don't think you need to specify that it's mapped to list, it's not that important implementation detail to user

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 4 above)
Comment (DanielChanger) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: Yeah, so it seems like we can avoid logging and do it in one stream pipeline
`scanners.stream().map(BeanScanner::scan).flatMap(List:Stream).toList()`
(but with newlines)

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 17]
```diff
@@ -1,60 +1,152 @@
...
+        List<BeanDefinition> beanDefinitionList = new ArrayList<>();
+        log.info("Scanning application for finding bean definitions");
+        for (BeanScanner beanScanner : scanners) {
+            log.debug("Running bean scanner {}", beanScanner.getClass().getSimpleName());
+            beanDefinitionList.addAll(beanScanner.scan());
+        }
+        return beanDefinitionList;
+    }
+
+    private void validate(BeanValidator validator, List<BeanDefinition> beanDefinitionList) {
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: (nit) not sure if this is worth extracting it to a method, because there's only one line in fact and log can be avoided here or at least changed to trace level

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 18]
```diff
@@ -1,60 +1,152 @@
...
+        log.info("Beans definitions validating");
+        validator.validate(beanDefinitionList);
+    }
+
+    private void contextCreation(List<BeanDefinition> beanDefinitionList) {
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: (nit) for me context creation sounds really generic, I'd rename it to something more specific like "mapToTable" or analogs

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 7 above)
Comment (DanielChanger) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:104]: Add here validation for blank, null and strings that contain spaces, please

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 8 above)
Comment (DanielChanger) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:116]: and doesn't have spaces anywhere

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 10 above)
Comment (DanielChanger) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:57]: and doesn't have any spaces

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+@TestInstance(value = TestInstance.Lifecycle.PER_CLASS)
+public class ApplicationContextImplTest {
+
+    private ApplicationContext applicationContext;
+    private BeanScanner beanScannerOne;
+    private BeanScanner beanScannerTwo;
+    private BeanValidator validator;
+    private BeanInitializer initializer;
+
+    @BeforeAll
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java:37]: The common practice is to make each test case independent, thus context has to be setup before each and not before all

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+import org.mockito.Mockito;
+
+import java.lang.reflect.Field;
+import java.math.BigDecimal;
+import java.util.*;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+@TestInstance(value = TestInstance.Lifecycle.PER_CLASS)
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java:28]: Why do we need it?

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+public class ApplicationContextImplTest {
+
+    private ApplicationContext applicationContext;
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java:31]: (nit) Consider using `@Mock` and `@InjectMock` here

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+
+        when(beanDefinitionStringTwo.name()).thenReturn("One more string bean");
+        doReturn(String.class).when(beanDefinitionStringTwo).type();
+        when(beanDefinitionStringTwo.instance()).thenReturn("String bean");
+
+        applicationContext = new ApplicationContextImpl(beanScannerList, validator, initializer);
+    }
+
+    @Test
+    @DisplayName("Application context initializing")
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: Add in this name what we expect

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+    public void applicationContextInitializingTest() throws NoSuchFieldException, IllegalAccessException {
+        verify(beanScannerOne).scan();
+        verify(beanScannerTwo).scan();
+
+        verify(validator).validate(anyList());
+
+        verify(initializer).initialize(any());
+
+        Field contextField = applicationContext.getClass().getDeclaredField("context");
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: I don't think we have to use reflection here, it's a bad practice. You can just leave verifies here and check the actual context size and values in other cases

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+    @Test
+    @DisplayName("Getting a bean from the context by bean type")
+    public void getBeanByType() {
+        Integer bean = applicationContext.getBean(Integer.class);
+        assertEquals(bean, 2);
+    }
+
+    @Test
+    @DisplayName("Throwing an exception when a bean with provided type is not found")
+    public void getBeanByTypeThrowsException1() {
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: Replace Exception1 with NoSuchBeanException

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+        assertThrows(NoSuchBeanException.class, () -> applicationContext.getBean(BigDecimal.class));
+    }
+
+    @Test
+    @DisplayName("Throwing an exception when more than one bean are found")
+    public void getBeanByTypeThrowsException2() {
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: Replace Exception1 with NoUniqueBeanException

**Review by DanielChanger — CHANGES_REQUESTED** [review]

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 7 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:104]: @DanielChanger do you mean that beanName should not contains spaces at all? Or it can be name with spaces but not blank?

Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:104]: I found the answer below :)

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 13 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: @NazarSmith228 I don't like "started", because when I see message like that than I expect message "ended" or "completed" appears. 
And one more thing - it's not started yet, it WILL be started.
What do you think about "Bean definition scanning" ?

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 16 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: @DanielChanger What do you think about this "Scanners scan application packages to define {@link BeanDefinition} configs." ?

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java (hunk 20 above)
Comment (nadya-gainutsa) [reply src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java:28]: @DanielChanger Tests methods don't change state, it's just getters. I didn't want to create application context for each method. This annotation allows one instance to be used for each methods.
If I remove this annotation, @BeforeAll annotation should be changed to @BeforeEach and annotated method init() to static. As a result, all class variables must to be static.

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 2 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: @NazarSmith228 Done

Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: @NazarSmith228 Done

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 3 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: Done

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 4 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: Done

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 5 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: Done

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 7 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:104]: Done

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 9 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:129]: Done

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 12 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: Done

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 14 above)
Comment (nadya-gainutsa) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: Done

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java (hunk 22 above)
Comment (nadya-gainutsa) [reply src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: @DanielChanger changed to "Scan, validate and initialized methods have been called". Is that ok? If you have a better idea, please share it with me.

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java (hunk 23 above)
Comment (nadya-gainutsa) [reply src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: @DanielChanger I'm not sure I understood you well. Do you propose to remove this code or to remove and add one more test method that will check size and values?

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java (hunk 24 above)
Comment (nadya-gainutsa) [reply src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: done

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java (hunk 25 above)
Comment (nadya-gainutsa) [reply src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: done

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java (hunk 13 above)
Comment (NazarSmith228) [reply src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: @nadya-gainutsa then let's log nothing here, since scanners will let us know that they started

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java (hunk 20 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java:28]: Okay, makes sense to me

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 26]
```diff
@@ -1,60 +1,152 @@
...
+
+        return scanners.stream()
+                .map(BeanScanner::scan)
+                .flatMap(List::stream)
+                .toList();
+    }
 
+    private Table<String, Class<?>, BeanDefinition> populateBeansDefinitionTable(List<BeanDefinition> beanDefinitionList) {
+        log.debug("Store scanned bean definition in application context");
+        HashBasedTable<String, Class<?>, BeanDefinition> beanDefinitionTable = HashBasedTable.create();
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: You should use either polymorphic reference `Table` or use `var` instead for simplicity

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 27]
```diff
@@ -1,60 +1,146 @@
...
-    public boolean containsBean(String name) {
-        return false;
+    public boolean containsBean(String beanName) {
+        checkBeanName(beanName);
+        return !beanDefinitionTable.row(beanName).isEmpty();
     }
+
+    private void checkBeanName(String beanName) {
+        checkArgument(isNotEmpty(beanName), "The argument [beanName] must be not null or empty");
+        checkArgument(isNotBlank(beanName), "The argument [beanName] must be not blank");
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java]: Hm, it seems to me that checking if we have any space in a name should be enough, so we don't need additional check for blank

File: src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+
+        when(beanDefinitionStringTwo.name()).thenReturn("One more string bean");
+        doReturn(String.class).when(beanDefinitionStringTwo).type();
+        when(beanDefinitionStringTwo.instance()).thenReturn("String bean");
+
+        applicationContext = new ApplicationContextImpl(beanScannerList, validator, initializer);
+    }
+
+    @Test
+    @DisplayName("Scan, validate and initialized methods have been called")
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/ApplicationContextImplTest.java]: (nit) I'd say "Calls bean scanners, context validator and bean initializer during on creation"

**Review by DanielChanger — CHANGES_REQUESTED** [review]

**Comment by DanielChanger** [comment]
@nadya-gainutsa please, fix sonar issues

**Review by NazarSmith228 — APPROVED** [review]

File: src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java [hunk 29]
```diff
@@ -1,60 +1,149 @@
...
+        checkNotNull(beanType, BEAN_TYPE_MUST_BE_NOT_NULL_MESSAGE);
+
+        List<T> beans = beanDefinitionTable.column(beanType)
+                .values()
+                .stream()
+                .map(beanDefinition -> beanType.cast(beanDefinition.instance()))
+                .toList();
+        if (beans.isEmpty()) {
+            throw new NoSuchBeanException(NO_SUCH_BEAN_EXCEPTION_MESSAGE.formatted(beanType.getSimpleName()));
+        } else if (beans.size() > 1) {
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java:98]: (nit)  else are not needed here

**Review by DanielChanger — APPROVED** [review]
LGTM
